### PR TITLE
gh-99377: Add audit events for thread creation and clear

### PR DESCRIPTION
**Pull request by zooba**
Also moves some asserts around to ensure we assert before using the things we're asserting

* Issue: gh-99377

### CODE REVIEW

File: Doc/c-api/init.rst [hunk 1]
```diff
@@ -1239,12 +1239,24 @@ All of the following functions must be called after :c:func:`Py_Initialize`.
    The global interpreter lock need not be held, but may be held if it is
    necessary to serialize calls to this function.
 
+   .. audit-event:: cpython.PyThreadState_New "" c.PyThreadState_New
+
+      Raise an auditing event ``cpython.PyThreadState_New`` with no arguments.
+      The event will be raised from the creating thread, not the new thread.
```
Comment (gpshead) [inline Doc/c-api/init.rst]: This wording might be confusing in some situations as isn't the creating thread the new thread in some circumstances such as when C/C++ has created a thread which is then calling back into the Python runtime?   Just stating "will be raised from the thread creating the new PyThreadState which may not be the new thread." would be more accurate in that case.

File: Modules/_threadmodule.c [hunk 2]
```diff
@@ -1145,6 +1145,11 @@ thread_PyThread_start_new_thread(PyObject *self, PyObject *fargs)
         return NULL;
     }
 
+    if (PySys_Audit("_thread.start_new_thread", "OOO",
```
Comment (gpshead) [inline Modules/_threadmodule.c:1148]: this one doesn't appear to be mentioned in the docs.

File: Python/sysmodule.c [hunk 3]
```diff
@@ -222,8 +222,8 @@ sys_audit_tstate(PyThreadState *ts, const char *event,
         PyDTrace_AUDIT(event, (void *)eventArgs);
     }
 
-    /* Call interpreter hooks */
-    if (is->audit_hooks) {
+    /* Call interpreter hooks if the GIL is held */
+    if (PyGILState_Check() && is->audit_hooks) {
```
Comment (gpshead) [inline Python/sysmodule.c]: is this related to the issue? it looks like a bugfix. best mention that in the logs/news if there was a problem before.

File: Modules/_threadmodule.c (hunk 2 above)
Comment (zooba) [reply Modules/_threadmodule.c:1148]: It's in the `_thread` docs, which is the most accurate but possibly the least helpful place to put it. But the main value of the docs is the table, so it'll appear there as the internal event that it is.

File: Python/sysmodule.c (hunk 3 above)
Comment (zooba) [reply Python/sysmodule.c]: It's related because we (apparently) never triggered any hooks from outside the GIL before. Now we potentially do, because PyGILState_Ensure may allocate a new threadstate, and while it _theoretically_ has the GIL, it isn't ready to call into any code. 

But I'm pretty sure nobody else could ever get into this state apart from us.

File: Doc/c-api/init.rst (hunk 1 above)
Comment (zooba) [reply Doc/c-api/init.rst]: I think PyGILState_Ensure is the edge case you're referring to, and historically it's changed behaviour so often that I'm hesitant to document anything about it. But I like your proposed wording.

File: Python/sysmodule.c (hunk 3 above)
Comment (zooba) [reply Python/sysmodule.c]: I think after my other changes we don't need this. Turns out, calling into any audit hooks at all with an invalid GIL is a bad idea 😅

**Review by gpshead — APPROVED** [review]
